File: taxfiler.client/src/app/model/document-match.ts
import { Document } from './document';

export interface DocumentMatch {
  document: Document;
  matchScore: number;
  scoreBreakdown: MatchScoreBreakdown;
}

/**
 * Detailed breakdown of match score components.
 * All scores are normalized to a range of 0.0 to 1.0, where:
 * - 0.0 represents no match
 * - 1.0 represents a perfect match
 */
export interface MatchScoreBreakdown {
  /** Score based on amount similarity (0.0 to 1.0) */
  amountScore: number;
  
  /** Score based on date proximity (0.0 to 1.0) */
  dateScore: number;
  
  /** Score based on vendor/counterparty matching (0.0 to 1.0) */
  vendorScore: number;
  
  /** Score based on reference number matching (0.0 to 1.0) */
  referenceScore: number;
  
  /** Final composite score after applying weights and bonuses (0.0 to 1.0) */
  compositeScore: number;
}

File: TaxFiler.Service/DocumentMatch.cs
using TaxFiler.DB.Model;

namespace TaxFiler.Service;

/// <summary>
/// Represents a document match result with confidence score and detailed breakdown.
/// </summary>
public class DocumentMatch
{
    /// <summary>
    /// The matched document.
    /// </summary>
    public Document Document { get; set; } = null!;
    
    /// <summary>
    /// Overall match confidence score (0.0 to 1.0).
    /// </summary>
    public double MatchScore { get; set; }
    
    /// <summary>
    /// Detailed breakdown of how the match score was calculated.
    /// </summary>
    public MatchScoreBreakdown ScoreBreakdown { get; set; } = null!;
}

/// <summary>
/// Detailed breakdown of match score components.
/// </summary>
public class MatchScoreBreakdown
{
    /// <summary>
    /// Score based on amount similarity (0.0 to 1.0).
    /// </summary>
    public double AmountScore { get; set; }
    
    /// <summary>
    /// Score based on date proximity (0.0 to 1.0).
    /// </summary>
    public double DateScore { get; set; }
    
    /// <summary>
    /// Score based on vendor/counterparty matching (0.0 to 1.0).
    /// </summary>
    public double VendorScore { get; set; }
    
    /// <summary>
    /// Score based on reference number matching (0.0 to 1.0).
    /// </summary>
    public double ReferenceScore { get; set; }
    
    /// <summary>
    /// Final composite score after applying weights and bonuses (0.0 to 1.0).
    /// </summary>
    public double CompositeScore { get; set; }
}